LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/Utils/TwinCollectionExtensions.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+namespace LoRaTools.Utils
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Globalization;
+    using LoRaWan;
+    using Microsoft.Azure.Devices.Shared;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+
+    public static class TwinCollectionExtensions
+    {
+        private static readonly Dictionary<Type, Func<object, object>> customConverters = new Dictionary<Type, Func<object, object>>
+        {
+            [typeof(StationEui)] = (s) => StationEui.Parse(s.ToString()),
+            [typeof(DevNonce)] = (s) => new DevNonce(Convert.ToUInt16(s, CultureInfo.InvariantCulture)),
+            [typeof(DevAddr)] = (s) => DevAddr.Parse(s.ToString()),
+        };
+
+        public static T? SafeRead<T>(this TwinCollection twinCollection, string property, T? defaultValue = default, ILogger? logger = null)
```

Comment:
[nit] `SafeRead<T>` and `ReadRequiredString` are currently only used from the `TwinCollectionReader`. We could move it there entirely, which would increase encapsulation a bit.

Reply:
The model is to expose everything through the extensions in the TwinCollection and have the reader as convenient to wrap the logging. I would like to keep that.